File: _stabileo_ref/engine/src/postprocess/diagrams_3d.rs
use crate::types::*;
use serde::{Deserialize, Serialize};
use std::collections::BTreeSet;

// ==================== Types ====================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DiagramPoint3D {
    pub t: f64,
    pub value: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ElementDiagram3D {
    pub element_id: usize,
    pub points: Vec<DiagramPoint3D>,
    pub max_value: f64,
    pub min_value: f64,
    pub max_abs_value: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DiagramResults3D {
    pub moment_y: Vec<ElementDiagram3D>,
    pub moment_z: Vec<ElementDiagram3D>,
    pub shear_y: Vec<ElementDiagram3D>,
    pub shear_z: Vec<ElementDiagram3D>,
    pub axial: Vec<ElementDiagram3D>,
    pub torsion: Vec<ElementDiagram3D>,
}

const NUM_POINTS: usize = 21;

// ==================== Sampling ====================

fn build_sampling_positions_3d(
    length: f64,
    point_loads: &[PointLoadInfo3D],
) -> Vec<f64> {
    let mut positions = BTreeSet::new();
    for i in 0..NUM_POINTS {
        let t = i as f64 / (NUM_POINTS - 1) as f64;
        positions.insert((t * 1e12) as i64);
    }
    let eps = 1e-6;
    for pl in point_loads {
        let t_pl = pl.a / length;
        if t_pl > eps { positions.insert(((t_pl - eps) * 1e12) as i64); }
        positions.insert((t_pl * 1e12) as i64);
        if t_pl < 1.0 - eps { positions.insert(((t_pl + eps) * 1e12) as i64); }
    }
    positions.into_iter().map(|k| k as f64 / 1e12).collect()
}

// ==================== Evaluate at t ====================

/// Evaluate a 3D diagram value at position t.
pub fn evaluate_diagram_3d_at(
    ef: &ElementForces3D,
    kind: &str,
    t: f64,
) -> f64 {
    let l = ef.length;
    let x = t * l;

    match kind {
        "momentZ" => {
            let mut value = ef.mz_start - ef.vy_start * x;
            for dl in &ef.distributed_loads_y {
                let a = dl.a;
                let b = dl.b;
                let span = b - a;
                if span < 1e-12 { continue; }
                let dq = dl.q_j - dl.q_i;
                let x_clamp = x.min(b);
                if x_clamp <= a { continue; }
                let s = x_clamp - a;
                value -= dl.q_i * (s * (x - a) - s * s / 2.0)
                       + dq / span * (s * s / 2.0 * (x - a) - s * s * s / 3.0);
            }
            for pl in &ef.point_loads_y {
                if pl.a < x - 1e-10 {
                    value -= pl.p * (x - pl.a);
                }
            }
            value
        }
        "momentY" => {
            let mut value = ef.my_start + ef.vz_start * x;
            for dl in &ef.distributed_loads_z {
                let a = dl.a;
                let b = dl.b;
                let span = b - a;
                if span < 1e-12 { continue; }
                let dq = dl.q_j - dl.q_i;
                let x_clamp = x.min(b);
                if x_clamp <= a { continue; }
                let s = x_clamp - a;
                value += dl.q_i * (s * (x - a) - s * s / 2.0)
                       + dq / span * (s * s / 2.0 * (x - a) - s * s * s / 3.0);
            }
            for pl in &ef.point_loads_z {
                if pl.a < x - 1e-10 {
                    value += pl.p * (x - pl.a);
                }
            }
            value
        }
        "shearY" => {
            let mut value = ef.vy_start;
            for dl in &ef.distributed_loads_y {
                let a = dl.a;
                let b = dl.b;
                let span = b - a;
                if span < 1e-12 { continue; }
                let dq = dl.q_j - dl.q_i;
                let x_clamp = x.min(b);
                if x_clamp <= a { continue; }
                let s = x_clamp - a;
                value += dl.q_i * s + dq * s * s / (2.0 * span);
            }
            for pl in &ef.point_loads_y {
                if pl.a < x - 1e-10 { value += pl.p; }
            }
            value
        }
        "shearZ" => {
            let mut value = ef.vz_start;
            for dl in &ef.distributed_loads_z {
                let a = dl.a;
                let b = dl.b;
                let span = b - a;
                if span < 1e-12 { continue; }
                let dq = dl.q_j - dl.q_i;
                let x_clamp = x.min(b);
                if x_clamp <= a { continue; }
                let s = x_clamp - a;
                value += dl.q_i * s + dq * s * s / (2.0 * span);
            }
            for pl in &ef.point_loads_z {
                if pl.a < x - 1e-10 { value += pl.p; }
            }
            value
        }
        "axial" => ef.n_start + t * (ef.n_end - ef.n_start),
        "torsion" => ef.mx_start + t * (ef.mx_end - ef.mx_start),
        _ => 0.0,
    }
}

// ==================== Compute Diagram ====================

fn compute_single_diagram_3d(
    ef: &ElementForces3D,
    kind: &str,
) -> ElementDiagram3D {
    let relevant_pl: &[PointLoadInfo3D] = match kind {
        "momentZ" | "shearY" => &ef.point_loads_y,
        "momentY" | "shearZ" => &ef.point_loads_z,
        _ => &[],
    };

    let positions = if relevant_pl.is_empty() {
        (0..NUM_POINTS).map(|i| i as f64 / (NUM_POINTS - 1) as f64).collect()
    } else {
        build_sampling_positions_3d(ef.length, relevant_pl)
    };

    let mut points = Vec::new();
    let mut max_val = f64::NEG_INFINITY;
    let mut min_val = f64::INFINITY;
    let mut max_abs_value: f64 = 0.0;

    for &t in &positions {
        let value = evaluate_diagram_3d_at(ef, kind, t);
        points.push(DiagramPoint3D { t, value });
        if value > max_val { max_val = value; }
        if value < min_val { min_val = value; }
        if value.abs() > max_abs_value.abs() {
            max_abs_value = value;
        }
    }

    ElementDiagram3D {
        element_id: ef.element_id,
        points,
        max_value: max_val,
        min_value: min_val,
        max_abs_value,
    }
}

/// Compute all 3D diagrams for all elements.
pub fn compute_diagrams_3d(results: &AnalysisResults3D) -> DiagramResults3D {
    let kinds = ["momentY", "momentZ", "shearY", "shearZ", "axial", "torsion"];
    // Iterate elements outermost so each element's forces stay in cache across
    // all six kinds, instead of six separate passes over the whole vector.
    let mut all: Vec<Vec<ElementDiagram3D>> =
        kinds.iter().map(|_| Vec::with_capacity(results.element_forces.len())).collect();
    for ef in &results.element_forces {
        for (k, kind) in kinds.iter().enumerate() {
            all[k].push(compute_single_diagram_3d(ef, kind));
        }
    }

    DiagramResults3D {
        moment_y: all.remove(0),
        moment_z: all.remove(0),
        shear_y: all.remove(0),
        shear_z: all.remove(0),
        axial: all.remove(0),
        torsion: all.remove(0),
    }
}

File: _stabileo_ref/web/src/lib/engine/diagrams-3d.ts
// 3D Diagram Computation — Analytical equations
// Compute diagram values (My, Mz, Vy, Vz, N, Mx) along each 3D element
// using equilibrium equations with distributed and point loads.

import type { ElementForces3D } from './types-3d';

export type Diagram3DKind = 'momentY' | 'momentZ' | 'shearY' | 'shearZ' | 'axial' | 'torsion';

export interface DiagramPoint3D {
  /** Normalized position along element [0, 1] */
  t: number;
  /** Diagram value at this point (kN or kN·m) */
  value: number;
}

export interface ElementDiagram3D {
  elementId: number;
  points: DiagramPoint3D[];
  maxValue: number;
  minValue: number;
  maxAbsValue: number;
}

const NUM_POINTS = 21;

/**
 * Build sorted, unique sampling positions (as t ∈ [0,1]) including
 * regular grid points and positions just before/after each point load
 * to capture discontinuities in shear diagrams.
 */
function buildSamplingPositions(
  length: number,
  pointLoads: Array<{ a: number; p: number }>,
): number[] {
  const tSet = new Set<number>();

  for (let i = 0; i < NUM_POINTS; i++) {
    tSet.add(i / (NUM_POINTS - 1));
  }

  const eps = 1e-6;
  for (const pl of pointLoads) {
    const tPl = pl.a / length;
    if (tPl > eps) tSet.add(tPl - eps);
    tSet.add(tPl);
    if (tPl < 1 - eps) tSet.add(tPl + eps);
  }

  return Array.from(tSet).sort((a, b) => a - b);
}

/**
 * Compute a 3D diagram for one element.
 * Delegates to evaluateDiagramAt for each sample point.
 */
export function computeDiagram3D(
  ef: ElementForces3D,
  kind: Diagram3DKind,
): ElementDiagram3D {
  const relevantPointLoads =
    kind === 'momentZ' || kind === 'shearY'
      ? ef.pointLoadsY
      : kind === 'momentY' || kind === 'shearZ'
        ? ef.pointLoadsZ
        : [];

  const positions = buildSamplingPositions(ef.length, relevantPointLoads);
  const points: DiagramPoint3D[] = [];
  let maxVal = -Infinity;
  let minVal = Infinity;
  let maxAbsValue = 0;

  for (const t of positions) {
    const value = evaluateDiagramAt(ef, kind, t);
    points.push({ t, value });
    if (value > maxVal) maxVal = value;
    if (value < minVal) minVal = value;
    if (Math.abs(value) > Math.abs(maxAbsValue)) maxAbsValue = value;
  }

  return { elementId: ef.elementId, points, maxValue: maxVal, minValue: minVal, maxAbsValue };
}

/**
 * Compute the global maximum absolute value for a diagram kind across all elements.
 */
export function computeGlobalMax3D(
  elementForces: ElementForces3D[],
  kind: Diagram3DKind,
): number {
  let globalMax = 0;
  for (const ef of elementForces) {
    const d = computeDiagram3D(ef, kind);
    globalMax = Math.max(globalMax, Math.abs(d.maxValue), Math.abs(d.minValue));
  }
  return globalMax;
}

/**
 * Determine the perpendicular direction for a diagram kind.
 *
 * - Mz, Vy  → bending in local XY plane  → perpendicular = ey  → return 'y'
 * - My, Vz  → bending in local XZ plane  → perpendicular = ez  → return 'z'
 * - N, Mx   → drawn in local XZ plane    → perpendicular = ez  → return 'z'
 */
export function getDiagramLocalDirection(kind: Diagram3DKind): 'y' | 'z' {
  switch (kind) {
    case 'momentZ':
    case 'shearY':
      return 'y';
    case 'momentY':
    case 'shearZ':
    case 'axial':
    case 'torsion':
      return 'z';
  }
}

/**
 * Evaluate the diagram value at a specific normalized position t ∈ [0,1].
 * Uses analytical equilibrium equations.
 */
export function evaluateDiagramAt(
  ef: ElementForces3D,
  kind: Diagram3DKind,
  t: number,
): number {
  const L = ef.length;
  const x = t * L;

  switch (kind) {
    case 'momentZ': {
      let value = ef.mzStart - ef.vyStart * x;
      for (const dl of ef.distributedLoadsY) {
        const a = dl.a;
        const b = dl.b;
        const span = b - a;
        if (span < 1e-12) continue;
        const dq = dl.qJ - dl.qI;
        const xClamp = Math.min(x, b);
        if (xClamp <= a) continue;
        const s = xClamp - a;
        value -= dl.qI * (s * (x - a) - s * s / 2)
               + dq / span * (s * s / 2 * (x - a) - s * s * s / 3);
      }
      for (const pl of ef.pointLoadsY) {
        if (pl.a < x - 1e-10) value -= pl.p * (x - pl.a);
      }
      return value;
    }
    case 'momentY': {
      let value = ef.myStart + ef.vzStart * x;
      for (const dl of ef.distributedLoadsZ) {
        const a = dl.a;
        const b = dl.b;
        const span = b - a;
        if (span < 1e-12) continue;
        const dq = dl.qJ - dl.qI;
        const xClamp = Math.min(x, b);
        if (xClamp <= a) continue;
        const s = xClamp - a;
        value += dl.qI * (s * (x - a) - s * s / 2)
               + dq / span * (s * s / 2 * (x - a) - s * s * s / 3);
      }
      for (const pl of ef.pointLoadsZ) {
        if (pl.a < x - 1e-10) value += pl.p * (x - pl.a);
      }
      return value;
    }
    case 'shearY': {
      let value = ef.vyStart;
      for (const dl of ef.distributedLoadsY) {
        const a = dl.a;
        const b = dl.b;
        const span = b - a;
        if (span < 1e-12) continue;
        const dq = dl.qJ - dl.qI;
        const xClamp = Math.min(x, b);
        if (xClamp <= a) continue;
        const s = xClamp - a;
        value += dl.qI * s + dq * s * s / (2 * span);
      }
      for (const pl of ef.pointLoadsY) {
        if (pl.a < x - 1e-10) value += pl.p;
      }
      return value;
    }
    case 'shearZ': {
      let value = ef.vzStart;
      for (const dl of ef.distributedLoadsZ) {
        const a = dl.a;
        const b = dl.b;
        const span = b - a;
        if (span < 1e-12) continue;
        const dq = dl.qJ - dl.qI;
        const xClamp = Math.min(x, b);
        if (xClamp <= a) continue;
        const s = xClamp - a;
        value += dl.qI * s + dq * s * s / (2 * span);
      }
      for (const pl of ef.pointLoadsZ) {
        if (pl.a < x - 1e-10) value += pl.p;
      }
      return value;
    }
    case 'axial':
      return ef.nStart + t * (ef.nEnd - ef.nStart);
    case 'torsion':
      return ef.mxStart + t * (ef.mxEnd - ef.mxStart);
  }
}

/** Format a 3D diagram value for display.
 *  Moment values are negated: internal convention is hogging=positive,
 *  but standard engineering convention is sagging=positive. */
export function formatDiagramValue3D(value: number, kind: Diagram3DKind): string {
  const isMoment = kind === 'momentY' || kind === 'momentZ' || kind === 'torsion';
  const displayVal = isMoment ? -value : value;
  const abs = Math.abs(displayVal);
  const sign = displayVal < 0 ? '-' : '';
  const formatted = abs >= 100 ? abs.toFixed(0) : abs >= 10 ? abs.toFixed(1) : abs.toFixed(2);
  const unit = isMoment ? ' kN·m' : ' kN';
  return sign + formatted + unit;
}
